File: backend/routers/channels.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession

from ..core.database import get_db
from ..models.server import Channel, ChannelType, Server, ChannelGroup
from .deps import CurrentUser, AdminUser
from ..core.permissions import check_channel_visibility, check_channel_speak_permission


router = APIRouter(prefix="/api/servers/{server_id}/channels", tags=["channels"])


class ChannelCreate(BaseModel):
    name: str
    type: str = "text"
    group_id: int | None = None  # Optional: assign to a channel group
    visibility_min_server_level: int = 1  # 1-4, default accessible to all
    visibility_min_internal_level: int = 1  # 1-2, default accessible to all
    speak_min_server_level: int = 1  # 1-4, default everyone can speak
    speak_min_internal_level: int = 1  # 1-2, default everyone can speak


class ChannelUpdate(BaseModel):
    name: str | None = None
    group_id: int | None = (
        None  # Optional: move to a different group (use -1 to ungroup)
    )
    visibility_min_server_level: int | None = None  # 1-4
    visibility_min_internal_level: int | None = None  # 1-2
    speak_min_server_level: int | None = None  # 1-4
    speak_min_internal_level: int | None = None  # 1-2


class ChannelResponse(BaseModel):
    id: int
    server_id: int
    group_id: int | None
    name: str
    type: str
    position: int
    top_position: int
    visibility_min_server_level: int = 1
    visibility_min_internal_level: int = 1
    speak_min_server_level: int = 1
    speak_min_internal_level: int = 1

    class Config:
        from_attributes = True


@router.get("", response_model=list[ChannelResponse])
async def list_channels(
    server_id: int, user: CurrentUser, db: AsyncSession = Depends(get_db)
):
    """List all channels in a server user has access to."""
    result = await db.execute(
        select(Channel).where(Channel.server_id == server_id).order_by(Channel.position)
    )
    channels = result.scalars().all()
    
    # Filter channels based on user's visibility permissions
    filtered_channels = [
        c for c in channels
        if check_channel_visibility(
            user,
            c.visibility_min_server_level,
            c.visibility_min_internal_level
        )
    ]
    
    return [
        ChannelResponse(
            id=c.id,
            server_id=c.server_id,
            group_id=c.group_id,
            name=c.name,
            type=c.type.value,
            position=c.position,
            top_position=c.top_position,
            visibility_min_server_level=c.visibility_min_server_level,
            visibility_min_internal_level=c.visibility_min_internal_level,
            speak_min_server_level=c.speak_min_server_level,
            speak_min_internal_level=c.speak_min_internal_level,
        )
        for c in filtered_channels
    ]


@router.post("", response_model=ChannelResponse, status_code=status.HTTP_201_CREATED)
async def create_channel(
    server_id: int,
    payload: ChannelCreate,
    user: AdminUser,
    db: AsyncSession = Depends(get_db),
):
    """Create a new channel (admin only)."""
    # Verify server exists
    server_result = await db.execute(select(Server).where(Server.id == server_id))
    if not server_result.scalar_one_or_none():
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Server not found"
        )

    # Verify group exists if provided
    if payload.group_id is not None:
        group_result = await db.execute(
            select(ChannelGroup).where(
                ChannelGroup.id == payload.group_id, ChannelGroup.server_id == server_id
            )
        )
        if not group_result.scalar_one_or_none():
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND, detail="Channel group not found"
            )

    channel_type = ChannelType.VOICE if payload.type == "voice" else ChannelType.TEXT
    
    # Validate permission levels
    if not (1 <= payload.visibility_min_server_level <= 4):
        raise HTTPException(status_code=400, detail="visibility_min_server_level must be 1-4")
    if not (1 <= payload.visibility_min_internal_level <= 2):
        raise HTTPException(status_code=400, detail="visibility_min_internal_level must be 1-2")
    if not (1 <= payload.speak_min_server_level <= 4):
        raise HTTPException(status_code=400, detail="speak_min_server_level must be 1-4")
    if not (1 <= payload.speak_min_internal_level <= 2):
        raise HTTPException(status_code=400, detail="speak_min_internal_level must be 1-2")

    if payload.group_id is not None:
        # Grouped channel: get max position within the group
        pos_result = await db.execute(
            select(func.max(Channel.position)).where(
                Channel.server_id == server_id, Channel.group_id == payload.group_id
            )
        )
        max_pos = pos_result.scalar() or -1
        channel = Channel(
            server_id=server_id,
            group_id=payload.group_id,
            name=payload.name,
            type=channel_type,
            position=max_pos + 1,
            top_position=0,  # Not used for grouped channels
            visibility_min_server_level=payload.visibility_min_server_level,
            visibility_min_internal_level=payload.visibility_min_internal_level,
            speak_min_server_level=payload.speak_min_server_level,
            speak_min_internal_level=payload.speak_min_internal_level,
        )
    else:
        # Ungrouped channel: get max top-level position
        group_max = await db.execute(
            select(func.max(ChannelGroup.position)).where(
                ChannelGroup.server_id == server_id
            )
        )
        channel_max = await db.execute(
            select(func.max(Channel.top_position)).where(
                Channel.server_id == server_id, Channel.group_id == None
            )
        )
        max_pos = max(group_max.scalar() or -1, channel_max.scalar() or -1)
        channel = Channel(
            server_id=server_id,
            group_id=None,
            name=payload.name,
            type=channel_type,
            position=0,  # Not used for ungrouped channels
            top_position=max_pos + 1,
            visibility_min_server_level=payload.visibility_min_server_level,
            visibility_min_internal_level=payload.visibility_min_internal_level,
            speak_min_server_level=payload.speak_min_server_level,
            speak_min_internal_level=payload.speak_min_internal_level,
        )

    db.add(channel)
    await db.flush()

    return ChannelResponse(
        id=channel.id,
        server_id=channel.server_id,
        group_id=channel.group_id,
        name=channel.name,
        type=channel.type.value,
        position=channel.position,
        top_position=channel.top_position,
        visibility_min_server_level=channel.visibility_min_server_level,
        visibility_min_internal_level=channel.visibility_min_internal_level,
        speak_min_server_level=channel.speak_min_server_level,
        speak_min_internal_level=channel.speak_min_internal_level,
    )


@router.patch("/{channel_id}", response_model=ChannelResponse)
async def update_channel(
    server_id: int,
    channel_id: int,
    payload: ChannelUpdate,
    user: AdminUser,
    db: AsyncSession = Depends(get_db),
):
    """Update channel properties (admin only)."""
    result = await db.execute(
        select(Channel).where(Channel.id == channel_id, Channel.server_id == server_id)
    )
    channel = result.scalar_one_or_none()
    if not channel:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Channel not found"
        )

    if payload.name is not None:
        channel.name = payload.name
    
    # Update permission levels
    if payload.visibility_min_server_level is not None:
        if not (1 <= payload.visibility_min_server_level <= 4):
            raise HTTPException(status_code=400, detail="visibility_min_server_level must be 1-4")
        channel.visibility_min_server_level = payload.visibility_min_server_level
    if payload.visibility_min_internal_level is not None:
        if not (1 <= payload.visibility_min_internal_level <= 2):
            raise HTTPException(status_code=400, detail="visibility_min_internal_level must be 1-2")
        channel.visibility_min_internal_level = payload.visibility_min_internal_level
    if payload.speak_min_server_level is not None:
        if not (1 <= payload.speak_min_server_level <= 4):
            raise HTTPException(status_code=400, detail="speak_min_server_level must be 1-4")
        channel.speak_min_server_level = payload.speak_min_server_level
    if payload.speak_min_internal_level is not None:
        if not (1 <= payload.speak_min_internal_level <= 2):
            raise HTTPException(status_code=400, detail="speak_min_internal_level must be 1-2")
        channel.speak_min_internal_level = payload.speak_min_internal_level

    # Handle group_id update: -1 means ungroup, None means no change, positive int means move to group
    if payload.group_id is not None:
        if payload.group_id == -1:
            # Ungroup the channel: assign top_position
            if channel.group_id is not None:
                group_max = await db.execute(
                    select(func.max(ChannelGroup.position)).where(
                        ChannelGroup.server_id == server_id
                    )
                )
                channel_max = await db.execute(
                    select(func.max(Channel.top_position)).where(
                        Channel.server_id == server_id, Channel.group_id == None
                    )
                )
                max_pos = max(group_max.scalar() or -1, channel_max.scalar() or -1)
                channel.group_id = None
                channel.top_position = max_pos + 1
                channel.position = 0
        else:
            # Move to a group: verify group exists
            group_result = await db.execute(
                select(ChannelGroup).where(
                    ChannelGroup.id == payload.group_id,
                    ChannelGroup.server_id == server_id,
                )
            )
            if not group_result.scalar_one_or_none():
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail="Channel group not found",
                )

            # Get max position in target group
            pos_result = await db.execute(
                select(func.max(Channel.position)).where(
                    Channel.server_id == server_id, Channel.group_id == payload.group_id
                )
            )
            max_pos = pos_result.scalar() or -1
            channel.group_id = payload.group_id
            channel.position = max_pos + 1
            channel.top_position = 0  # Not used for grouped channels

    await db.flush()

    return ChannelResponse(
        id=channel.id,
        server_id=channel.server_id,
        group_id=channel.group_id,
        name=channel.name,
        type=channel.type.value,
        position=channel.position,
        top_position=channel.top_position,
        visibility_min_server_level=channel.visibility_min_server_level,
        visibility_min_internal_level=channel.visibility_min_internal_level,
        speak_min_server_level=channel.speak_min_server_level,
        speak_min_internal_level=channel.speak_min_internal_level,
    )


@router.delete("/{channel_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_channel(
    server_id: int, channel_id: int, user: AdminUser, db: AsyncSession = Depends(get_db)
):
    """Delete a channel (admin only)."""
    result = await db.execute(
        select(Channel).where(Channel.id == channel_id, Channel.server_id == server_id)
    )
    channel = result.scalar_one_or_none()
    if not channel:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Channel not found"
        )

    await db.delete(channel)

File: backend-go/internal/handler/channels.go
package handler

import (
	"database/sql"
	"net/http"
	"strconv"

	"github.com/labstack/echo/v4"

	"github.com/RMS-Server/rms-discord-go/internal/middleware"
	"github.com/RMS-Server/rms-discord-go/internal/permission"
	"github.com/RMS-Server/rms-discord-go/internal/sso"
)

// ChannelHandler handles channel CRUD endpoints.
type ChannelHandler struct {
	db  *sql.DB
	sso *sso.Client
}

func NewChannelHandler(db *sql.DB, sso *sso.Client) *ChannelHandler {
	return &ChannelHandler{db: db, sso: sso}
}

type channelCreateReq struct {
	Name               string `json:"name"`
	Type               string `json:"type"`
	GroupID            *int64 `json:"group_id"`
	MinLevel           int    `json:"min_level"`
	SpeakMinLevel      int    `json:"speak_min_level"`
	PermMinLevel       int    `json:"perm_min_level"`
	LogicOperator      string `json:"logic_operator"`
	SpeakPermMinLevel  int    `json:"speak_perm_min_level"`
	SpeakLogicOperator string `json:"speak_logic_operator"`
}

type channelUpdateReq struct {
	Name               *string `json:"name"`
	GroupID            *int64  `json:"group_id"`
	MinLevel           *int    `json:"min_level"`
	SpeakMinLevel      *int    `json:"speak_min_level"`
	PermMinLevel       *int    `json:"perm_min_level"`
	LogicOperator      *string `json:"logic_operator"`
	SpeakPermMinLevel  *int    `json:"speak_perm_min_level"`
	SpeakLogicOperator *string `json:"speak_logic_operator"`
}

type channelResponse struct {
	ID                 int64  `json:"id"`
	ServerID           int64  `json:"server_id"`
	GroupID            *int64 `json:"group_id"`
	Name               string `json:"name"`
	Type               string `json:"type"`
	Position           int    `json:"position"`
	TopPosition        int    `json:"top_position"`
	MinLevel           int    `json:"min_level"`
	SpeakMinLevel      int    `json:"speak_min_level"`
	PermMinLevel       int    `json:"perm_min_level"`
	LogicOperator      string `json:"logic_operator"`
	SpeakPermMinLevel  int    `json:"speak_perm_min_level"`
	SpeakLogicOperator string `json:"speak_logic_operator"`
}

// ListChannels returns channels filtered by user visibility.
// GET /api/servers/:server_id/channels
func (h *ChannelHandler) ListChannels(c echo.Context) error {
	user := middleware.GetUser(c)
	serverID, err := strconv.ParseInt(c.Param("server_id"), 10, 64)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid server id"})
	}

	rows, err := h.db.Query(
		`SELECT id, server_id, group_id, name, type, position, top_position,
		        min_level, speak_min_level,
		        perm_min_level, logic_operator, speak_perm_min_level, speak_logic_operator
		 FROM channels WHERE server_id = ? ORDER BY position`, serverID,
	)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	defer rows.Close()

	var channels []channelResponse
	for rows.Next() {
		var ch channelResponse
		if err := rows.Scan(&ch.ID, &ch.ServerID, &ch.GroupID, &ch.Name, &ch.Type,
			&ch.Position, &ch.TopPosition,
			&ch.MinLevel, &ch.SpeakMinLevel,
			&ch.PermMinLevel, &ch.LogicOperator, &ch.SpeakPermMinLevel, &ch.SpeakLogicOperator); err != nil {
			return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
		}
		rule := permission.PermRule{PermMinLevel: ch.PermMinLevel, GroupMinLevel: ch.MinLevel, LogicOperator: ch.LogicOperator}
		if permission.CanAccess(user, rule) {
			channels = append(channels, ch)
		}
	}
	if channels == nil {
		channels = []channelResponse{}
	}
	return c.JSON(http.StatusOK, channels)
}

// CreateChannel creates a new channel in a server.
// POST /api/servers/:server_id/channels
func (h *ChannelHandler) CreateChannel(c echo.Context) error {
	serverID, err := strconv.ParseInt(c.Param("server_id"), 10, 64)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid server id"})
	}

	var req channelCreateReq
	if err := c.Bind(&req); err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid request body"})
	}
	if req.Name == "" {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "name is required"})
	}
	if req.Type == "" {
		req.Type = "text"
	}

	// Validate permission levels
	if req.MinLevel < 0 {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "min_level must be >= 0"})
	}
	if req.SpeakMinLevel < 0 {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "speak_min_level must be >= 0"})
	}
	if req.PermMinLevel < 0 {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "perm_min_level must be >= 0"})
	}
	if req.SpeakPermMinLevel < 0 {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "speak_perm_min_level must be >= 0"})
	}

	// Default logic operators
	if req.LogicOperator == "" {
		req.LogicOperator = "AND"
	}
	if req.SpeakLogicOperator == "" {
		req.SpeakLogicOperator = "AND"
	}
	if req.LogicOperator != "AND" && req.LogicOperator != "OR" {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "logic_operator must be AND or OR"})
	}
	if req.SpeakLogicOperator != "AND" && req.SpeakLogicOperator != "OR" {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "speak_logic_operator must be AND or OR"})
	}

	// Verify server exists
	var exists int
	if err := h.db.QueryRow("SELECT 1 FROM servers WHERE id = ?", serverID).Scan(&exists); err != nil {
		return c.JSON(http.StatusNotFound, map[string]string{"error": "server not found"})
	}

	channelType := "TEXT"
	if req.Type == "voice" {
		channelType = "VOICE"
	}

	var position, topPosition int
	if req.GroupID != nil {
		// Verify group exists
		if err := h.db.QueryRow("SELECT 1 FROM channel_groups WHERE id = ? AND server_id = ?", *req.GroupID, serverID).Scan(&exists); err != nil {
			return c.JSON(http.StatusNotFound, map[string]string{"error": "channel group not found"})
		}
		// Get max position within group
		var maxPos sql.NullInt64
		h.db.QueryRow("SELECT MAX(position) FROM channels WHERE server_id = ? AND group_id = ?", serverID, *req.GroupID).Scan(&maxPos)
		if maxPos.Valid {
			position = int(maxPos.Int64) + 1
		}
	} else {
		// Ungrouped: get max top-level position
		var groupMax, chMax sql.NullInt64
		h.db.QueryRow("SELECT MAX(position) FROM channel_groups WHERE server_id = ?", serverID).Scan(&groupMax)
		h.db.QueryRow("SELECT MAX(top_position) FROM channels WHERE server_id = ? AND group_id IS NULL", serverID).Scan(&chMax)
		gm, cm := int64(-1), int64(-1)
		if groupMax.Valid {
			gm = groupMax.Int64
		}
		if chMax.Valid {
			cm = chMax.Int64
		}
		if gm > cm {
			topPosition = int(gm) + 1
		} else {
			topPosition = int(cm) + 1
		}
	}

	res, err := h.db.Exec(
		`INSERT INTO channels (server_id, group_id, name, type, position, top_position,
		    min_level, speak_min_level,
		    perm_min_level, logic_operator, speak_perm_min_level, speak_logic_operator)
		 VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		serverID, req.GroupID, req.Name, channelType, position, topPosition,
		req.MinLevel, req.SpeakMinLevel,
		req.PermMinLevel, req.LogicOperator, req.SpeakPermMinLevel, req.SpeakLogicOperator,
	)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	chID, _ := res.LastInsertId()

	return c.JSON(http.StatusCreated, channelResponse{
		ID: chID, ServerID: serverID, GroupID: req.GroupID,
		Name: req.Name, Type: channelType,
		Position: position, TopPosition: topPosition,
		MinLevel: req.MinLevel, SpeakMinLevel: req.SpeakMinLevel,
		PermMinLevel: req.PermMinLevel, LogicOperator: req.LogicOperator,
		SpeakPermMinLevel: req.SpeakPermMinLevel, SpeakLogicOperator: req.SpeakLogicOperator,
	})
}

// UpdateChannel updates channel properties.
// PATCH /api/channels/:id
func (h *ChannelHandler) UpdateChannel(c echo.Context) error {
	chID, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid channel id"})
	}

	// Fetch current channel
	var ch channelResponse
	err = h.db.QueryRow(
		`SELECT id, server_id, group_id, name, type, position, top_position,
		        min_level, speak_min_level,
		        perm_min_level, logic_operator, speak_perm_min_level, speak_logic_operator
		 FROM channels WHERE id = ?`, chID,
	).Scan(&ch.ID, &ch.ServerID, &ch.GroupID, &ch.Name, &ch.Type,
		&ch.Position, &ch.TopPosition,
		&ch.MinLevel, &ch.SpeakMinLevel,
		&ch.PermMinLevel, &ch.LogicOperator, &ch.SpeakPermMinLevel, &ch.SpeakLogicOperator)
	if err == sql.ErrNoRows {
		return c.JSON(http.StatusNotFound, map[string]string{"error": "channel not found"})
	}
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}

	var req channelUpdateReq
	if err := c.Bind(&req); err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid request body"})
	}

	if req.Name != nil {
		ch.Name = *req.Name
	}
	if req.MinLevel != nil {
		if *req.MinLevel < 0 {
			return c.JSON(http.StatusBadRequest, map[string]string{"error": "min_level must be >= 0"})
		}
		ch.MinLevel = *req.MinLevel
	}
	if req.SpeakMinLevel != nil {
		if *req.SpeakMinLevel < 0 {
			return c.JSON(http.StatusBadRequest, map[string]string{"error": "speak_min_level must be >= 0"})
		}
		ch.SpeakMinLevel = *req.SpeakMinLevel
	}
	if req.PermMinLevel != nil {
		if *req.PermMinLevel < 0 {
			return c.JSON(http.StatusBadRequest, map[string]string{"error": "perm_min_level must be >= 0"})
		}
		ch.PermMinLevel = *req.PermMinLevel
	}
	if req.LogicOperator != nil {
		if *req.LogicOperator != "AND" && *req.LogicOperator != "OR" {
			return c.JSON(http.StatusBadRequest, map[string]string{"error": "logic_operator must be AND or OR"})
		}
		ch.LogicOperator = *req.LogicOperator
	}
	if req.SpeakPermMinLevel != nil {
		if *req.SpeakPermMinLevel < 0 {
			return c.JSON(http.StatusBadRequest, map[string]string{"error": "speak_perm_min_level must be >= 0"})
		}
		ch.SpeakPermMinLevel = *req.SpeakPermMinLevel
	}
	if req.SpeakLogicOperator != nil {
		if *req.SpeakLogicOperator != "AND" && *req.SpeakLogicOperator != "OR" {
			return c.JSON(http.StatusBadRequest, map[string]string{"error": "speak_logic_operator must be AND or OR"})
		}
		ch.SpeakLogicOperator = *req.SpeakLogicOperator
	}

	// Handle group_id change: -1 means ungroup
	if req.GroupID != nil {
		if *req.GroupID == -1 {
			// Ungroup: assign top_position
			if ch.GroupID != nil {
				var groupMax, chMax sql.NullInt64
				h.db.QueryRow("SELECT MAX(position) FROM channel_groups WHERE server_id = ?", ch.ServerID).Scan(&groupMax)
				h.db.QueryRow("SELECT MAX(top_position) FROM channels WHERE server_id = ? AND group_id IS NULL", ch.ServerID).Scan(&chMax)
				gm, cm := int64(-1), int64(-1)
				if groupMax.Valid {
					gm = groupMax.Int64
				}
				if chMax.Valid {
					cm = chMax.Int64
				}
				maxPos := gm
				if cm > maxPos {
					maxPos = cm
				}
				ch.GroupID = nil
				ch.TopPosition = int(maxPos) + 1
				ch.Position = 0
			}
		} else {
			// Move to group
			var exists int
			if err := h.db.QueryRow("SELECT 1 FROM channel_groups WHERE id = ? AND server_id = ?", *req.GroupID, ch.ServerID).Scan(&exists); err != nil {
				return c.JSON(http.StatusNotFound, map[string]string{"error": "channel group not found"})
			}
			var maxPos sql.NullInt64
			h.db.QueryRow("SELECT MAX(position) FROM channels WHERE server_id = ? AND group_id = ?", ch.ServerID, *req.GroupID).Scan(&maxPos)
			pos := 0
			if maxPos.Valid {
				pos = int(maxPos.Int64) + 1
			}
			ch.GroupID = req.GroupID
			ch.Position = pos
			ch.TopPosition = 0
		}
	}

	_, err = h.db.Exec(
		`UPDATE channels SET name = ?, group_id = ?, position = ?, top_position = ?,
		    min_level = ?, speak_min_level = ?,
		    perm_min_level = ?, logic_operator = ?, speak_perm_min_level = ?, speak_logic_operator = ?
		 WHERE id = ?`,
		ch.Name, ch.GroupID, ch.Position, ch.TopPosition,
		ch.MinLevel, ch.SpeakMinLevel,
		ch.PermMinLevel, ch.LogicOperator, ch.SpeakPermMinLevel, ch.SpeakLogicOperator, chID,
	)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}

	return c.JSON(http.StatusOK, ch)
}

// DeleteChannel deletes a channel.
// DELETE /api/channels/:id
func (h *ChannelHandler) DeleteChannel(c echo.Context) error {
	chID, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid channel id"})
	}

	res, err := h.db.Exec("DELETE FROM channels WHERE id = ?", chID)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	n, _ := res.RowsAffected()
	if n == 0 {
		return c.JSON(http.StatusNotFound, map[string]string{"error": "channel not found"})
	}
	return c.NoContent(http.StatusNoContent)
}
